testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestHiveBucketedTables.java

```diff
@@ -289,16 +299,98 @@ public void testBucketingVersion()
+        testBucketingVersion(BUCKETED_DEFAULT, value, true, (getHiveVersionMajor() < 3) ? bucketV1NameOptions : bucketV2NameOptions);
+        testBucketingVersion(BUCKETED_V1, value, false, bucketV1NameOptions);
+        testBucketingVersion(BUCKETED_V1, value, true, bucketV1NameOptions);
         if (getHiveVersionMajor() >= 3) {
             testBucketingVersion(BucketingType.BUCKETED_V2, value, false, bucketV2NameOptions);
             testBucketingVersion(BucketingType.BUCKETED_V2, value, true, bucketV2NameOptions);
         }
     }
 
+    @Test(dataProvider = "testBucketingWithUnsupportedDataTypesDataProvider")
+    public void testBucketingWithUnsupportedDataTypes(BucketingType bucketingType, String columnToBeBucketed)
+    {
+        try (TemporaryHiveTable table = temporaryHiveTable("table_with_unsupported_bucketing_types_" + randomTableSuffix())) {
+            String tableName = table.getName();
+            onHive().executeQuery(format("CREATE TABLE %s (" +
+                            "n_integer       INT," +
+                            "n_decimal       DECIMAL(9, 2)," +
+                            "n_timestamp     TIMESTAMP," +
+                            "n_char          CHAR(10)," +
+                            "n_binary        BINARY," +
+                            "n_union         UNIONTYPE<INT,STRING>," +
+                            "n_struct        STRUCT<field1:INT,field2:STRING>) " +
+                            "CLUSTERED BY (%s) INTO 2 BUCKETS " +
+                            "STORED AS ORC " +
+                            "%s",
+                    tableName,
+                    columnToBeBucketed,
+                    hiveTableProperties(bucketingType)));
+
+            populateRowToHiveTable(
+                    tableName,
+                    ImmutableList.<String>builder()
+                            .add("1")
+                            .add("CAST(1 AS DECIMAL(9, 2))")
+                            .add("current_timestamp()")
+                            .add("'Trino'")
+                            .add("binary('Trino')")
+                            .add("create_union(0, 1, 'Trino')")
+                            .add("named_struct('field1', 1, 'field2', 'Field2')")
+                            .build(),
+                    Optional.empty());
+
+            assertThat(onTrino().executeQuery(format("SELECT * FROM %s", tableName)))
+                    .hasRowsCount(1);
+
+            assertQueryFailure(() -> onTrino().executeQuery("SELECT \"$bucket\" FROM " + tableName))
+                    .hasMessageContaining("Column '$bucket' cannot be resolved");
+
+            assertQueryFailure(() -> onTrino().executeQuery(format("INSERT INTO %s(n_integer) VALUES (1)", tableName)))
+                    .hasMessageContaining("Cannot write to a table bucketed on an unsupported type");
```

Comment:
can we include the problematic type in the message?

```suggestion
                    .hasMessage("Cannot write to a table bucketed on an unsupported type: struct");
```

Reply:
It would be a bit tricky. So skipping this.